src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -2097,23 +2065,44 @@ protected override void AfterLeftChildHasBeenVisited(BoundBinaryOperator binary)
 
                     // `x == null` and `x != null` are pure null tests so update the null-state in the alternative branch too
```

Comment:
>// `x == null` and `x != null` are pure null tests so update the null-state in the alternative branch too [](start = 20, length = 105)

Reading this, I realized we probably have a bug: `x == nullConstant` is going to set the state of `x` to maybe-null, but it should not.
Could you add a test and either fix or file a bug? Only the `null` literal should count as a pure null test.